File: solutions/csharp/protein-translation/2/ProteinTranslation.cs
using System;
using System.Linq;

public static class ProteinTranslation
{
    public static string[] Proteins(string strand)
        => strand.Chunk(3)
            .Select(chars => new string(chars))
            .TakeWhile(protein => !IsStop(protein))
            .Select(Translate)
            .ToArray();

    private static bool IsStop(string s)
        => s is "UAA" or "UAG" or "UGA";

    private static string Translate(string protein)
        => protein switch
        {
            "AUG" => "Methionine",
            "UUU" or "UUC" => "Phenylalanine",
            "UUA" or "UUG" => "Leucine",
            "UCU" or "UCC" or "UCA" or "UCG" => "Serine",
            "UAU" or "UAC" => "Tyrosine",
            "UGU" or "UGC" => "Cysteine",
            "UGG" => "Tryptophan",
            _ => throw new InvalidOperationException("Invalid protein")
        };
}

File: solutions/fsharp/protein-translation/3/ProteinTranslation.fs
module ProteinTranslation

let proteins (rna:string) = 
    let isStop = function
        | "UAA" | "UAG" | "UGA" -> true
        | _ -> false

    let translate = function
        | "AUG" -> "Methionine"
        | "UUU" | "UUC" -> "Phenylalanine" 
        | "UUA" | "UUG" -> "Leucine"
        | "UCU" | "UCC" | "UCA" | "UCG" -> "Serine"
        | "UAU" | "UAC" -> "Tyrosine"
        | "UGU" | "UGC" -> "Cysteine"
        | "UGG" -> "Tryptophan"
        | _ -> ""

    rna 
    |> Seq.chunkBySize 3
    |> Seq.map System.String
    |> Seq.takeWhile (isStop >> not)
    |> Seq.map(translate)
    |> Seq.toList

